src/XMakeTasks/UnitTests/Copy_Tests.cs

```diff
@@ -2120,4 +2161,83 @@ public void CopyToDestinationFolderWithHardLinkFallbackTooManyLinks()
+            string destFile = Path.Combine(destFolder, Path.GetFileName(sourceFile));
+            try
+            {
+                using (StreamWriter sw = new StreamWriter(sourceFile, true))    // HIGHCHAR: Test writes in UTF8 without preamble.
+                    sw.Write("This is a source temp file.");
+
+                // Don't create the dest folder, let task do that
+
+                ITaskItem[] sourceFiles = new ITaskItem[] { new TaskItem(sourceFile) };
+
+                Copy t = new Copy();
+                t.RetryDelayMilliseconds = 1; // speed up tests!
+
+                // Allow the task's default (false) to have a chance
+                t.UseSymboliclinksIfPossible = true;
+
+                MockEngine me = new MockEngine(true);
+                t.BuildEngine = me;
+                t.SourceFiles = sourceFiles;
+                t.DestinationFolder = new TaskItem(destFolder);
+                t.SkipUnchangedFiles = true;
+
+                bool success = t.Execute();
+
+                Assert.True(success); // "success"
+                Assert.True(File.Exists(destFile)); // "destination exists"
+                Microsoft.Build.UnitTests.MockEngine.GetStringDelegate resourceDelegate = new Microsoft.Build.UnitTests.MockEngine.GetStringDelegate(AssemblyResources.GetString);
+
+                me.AssertLogContainsMessageFromResource(resourceDelegate, "Copy.SymbolicLinkComment", sourceFile, destFile);
+
+                string destinationFileContents;
+                using (StreamReader sr = new StreamReader(destFile))
+                    destinationFileContents = sr.ReadToEnd();
+
+                Assert.Equal(destinationFileContents, "This is a source temp file."); //                     "Expected the destination symbolic linked file to contain the contents of source file."
+
+                Assert.Equal(1, t.DestinationFiles.Length);
+                Assert.Equal(1, t.CopiedFiles.Length);
+                Assert.Equal(destFile, t.DestinationFiles[0].ItemSpec);
+                Assert.Equal(destFile, t.CopiedFiles[0].ItemSpec);
+
+                // Now we will write new content to the source file
+                // we'll then check that the destination file automatically
+                // has the same content (i.e. it's been hard linked)
+                using (StreamWriter sw = new StreamWriter(sourceFile, false))    // HIGHCHAR: Test writes in UTF8 without preamble.
+                    sw.Write("This is another source temp file.");
+
+                // Read the destination file (it should have the same modified content as the source)
+                using (StreamReader sr = new StreamReader(destFile))
+                    destinationFileContents = sr.ReadToEnd();
+
+                Assert.Equal(destinationFileContents, "This is another source temp file."); //                     "Expected the destination hard linked file to contain the contents of source file. Even after modification of the source"
```

Comment:
While you're in here, can you add a commit that swaps the order of these arguments? It should be `expected, actual`.  Please do so above in the hardlink test as well.